lib/demo/accounts/user.ex

```diff
@@ -0,0 +1,119 @@
+defmodule Demo.Accounts.User do
+  use Ecto.Schema
+  import Ecto.Changeset
+
+  schema "users" do
+    field :email, :string
+    field :password, :string, virtual: true
+    field :hashed_password, :string
+    field :confirmed_at, :naive_datetime
+
+    timestamps()
+  end
+
+  @doc """
+  A user changeset for registration.
+
+  It is important to validate the length of both e-mail and password.
+  Otherwise databases may truncate them without warnings, which could
+  lead to unpredictable or insecure behaviour. Long passwords may also
+  be very expensive to hash.
+  """
+  def registration_changeset(user, attrs) do
+    user
+    |> cast(attrs, [:email, :password])
+    |> validate_email()
+    |> validate_password()
```

Comment:
Should the `:password` change be deleted from the changeset after is used in password hashing? That way the user returned from `Ecto.Repo.insert/1` will have the `:password` field blanked out, and there is less risk of the plaintext password ending up anywhere (logs, process state, etc) if anything else is done with the returned user.

This would have the usability drawback of requiring that the user re-enter their password if the changeset is invalid and they're sent back to the registration form.

Since the same argument applies to password changes, if this is a valid concern, then perhaps `maybe_hash_password` would be the correct place to do this?

Reply:
If the concern is leaking data, I believe a decent solution is: `@derive {Inspect, except: [:password]}`, it will obviously only help where inspect is used.

Reply:
That's a great idea, and would partially remedy the problem.

One of the situations I'm imagining is someone building a write-through cache (maybe using Cachex?), and tossing the newly created user into it. The cache would then contain plaintext passwords after a user is created, until the cache is flushed.

Reply:
maybe blanking out the :password field at the last step of the pipe and only if the changeset is valid can fix this concern.